### DISCUSSION TITLE
flattenObject

### DISCUSSION THREAD

**Discussion by aturancetin** [remeda/remeda#971] [Suggest a Function]
Hi! Do you think it would be useful to have a function to flatten objects by recursively iterating through? What I mean is ;

```
var unshapedObj = { a: 'a', b: { ba: { baa: 'baa', bab: 'bab' } } }
var flattenedObj = R.flatObj(unshapedObj)
console.log(flattenedObj)
// => { a: 'a', 'b.ba.baa': 'baa', 'b.ba.bab': 'bab' }
```

**Comment by TkDodo** [comment]
not sure this would be useful to be honest. Why would we want to have the keys as dot notation ?

**Comment by aturancetin** [comment]
We use this functionality in our project to get nested keys to display some data from an array. Maybe we do overengineering here i am not sure :) , but if it would be interesting for someone i would try to create a PR about this but I also thought since we are mutating the object in this context, maybe it would not be a good fit for a typescript library.

**Comment by vlad-iakovlev** [comment]
It would be almost impossible to cover this function with auto-generated types

**Comment by silveltman** [comment]
Please have a loot at this, works very well with good typescript support: https://www.npmjs.com/package/flatten-anything

**Comment by cjquines** [comment]
i still don't see a good usecase here explained here (the one in [this comment](https://github.com/remeda/remeda/issues/170#issuecomment-1105858954) feels oddly specific, and i'd like to hear of other use cases)

**Comment by silveltman** [comment]
Mostly to do one of the following things without needing recursive functions:
- Apply a transformation to all keys of a deeply nested object
- Apply a transformation to all values of a deeply nested object 

Some more specific examples:
- Change casing
- Replace template vars in values
- Remove specific values, like nulls

Some super specific examples:
- I have a key names '_' which i use as kind of a preset/defaults/template key. When I set this to an object I want it destructured so it merges with its parent. With a flattened object I can simply map the object keys and remove that part of the string
- I want to mergeAndConcat an object (also references [here](https://github.com/remeda/remeda/issues/657) as seperate function). Having a flat object would allow me to do this by replacing the number in the key (mykey[5].otherkey) with an index of the mapKeys loop. Then you know for sure all arrays items at the same level have a unique index and they go up the further in the object you go, therefore concatting instead of overwriting

**Comment by cjquines** [comment]
hm, it's true that remeda isn't great for working with deeply nested objects, but i think we can avoid using path-based tools for this. in this case, it sounds like a `mapKeysDeep` and `mapValuesDeep` would suffice for the first few cases you mention

> I have a key names '_' which i use as kind of a preset/defaults/template key. When I set this to an object I want it destructured so it merges with its parent.

compare
```ts
({ _, ...rest }) => mergeDeep(_, rest);
```

with
```ts
(data) => pipe(
  data,
  flatten(),
  mapKeys(key => key.startsWith('_.') ? key.slice(2) : key),
  nestify(),
);
```

(the corresponding deep version with `mapValuesDeep` also isn't too bad, i think)

> I want to mergeAndConcat an object

in this case we should have a utility for that, rather than working with keys...

**Comment by Brodzko** [comment]
> hm, it's true that remeda isn't great for working with deeply nested objects

Just chiming in here, is this something you would _like_ to see resolved (v2, v3 even)? Or are you trying to avoid it because typing support for path-based utilities is difficult to get right (it totally is)? 

It does feel like a bit of a nuisance having to work around this limitation when we can't avoid deeply nested objects in our codebases.

I would not mind at all having some sort of optics utilities available as part of Remeda (or even as a sub-module of Remeda if that's preferred)

**Comment by cjquines** [comment]
my point was that path-based utilities are not the only approach to deeply nested objects; i even suggested two utilities that could help with this (`mapKeysDeep`, `mapValuesDeep`).

**Comment by silveltman** [comment]
I agree, mapKeysDeep and mapValuesDeep are both good options too
